builder/relay.go

```diff
@@ -0,0 +1,218 @@
+package builder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	boostTypes "github.com/flashbots/go-boost-utils/types"
+	"github.com/flashbots/mev-boost/server"
+)
+
+type testRelay struct {
+	validator     ValidatorData
+	requestedSlot uint64
+	submittedMsg  *boostTypes.BuilderSubmitBlockRequest
+}
+
+func (r *testRelay) SubmitBlock(msg *boostTypes.BuilderSubmitBlockRequest) error {
+	r.submittedMsg = msg
+	return nil
+}
+func (r *testRelay) GetValidatorForSlot(nextSlot uint64) (ValidatorData, error) {
+	r.requestedSlot = nextSlot
+	return r.validator, nil
+}
+func (r *testRelay) handleRegisterValidator(w http.ResponseWriter, req *http.Request) {
+}
+
+type RemoteRelay struct {
+	endpoint string
+	client   http.Client
+
+	localRelay *LocalRelay
+
+	validatorsLock       sync.RWMutex
+	validatorSyncOngoing bool
+	lastRequestedSlot    uint64
+	validatorSlotMap     map[uint64]ValidatorData
+}
+
+func NewRemoteRelay(endpoint string, localRelay *LocalRelay) (*RemoteRelay, error) {
+	r := &RemoteRelay{
+		endpoint:             endpoint,
+		client:               http.Client{Timeout: time.Second},
+		localRelay:           localRelay,
+		validatorSyncOngoing: false,
+		lastRequestedSlot:    0,
+		validatorSlotMap:     make(map[uint64]ValidatorData),
+	}
+
+	err := r.updateValidatorsMap(0, 3)
+	return r, err
+}
+
+type GetValidatorRelayResponse []struct {
+	Slot  string `json:"slot"`
+	Entry struct {
+		Message struct {
+			FeeRecipient string `json:"fee_recipient"`
+			GasLimit     string `json:"gas_limit"`
+			Timestamp    string `json:"timestamp"`
+			Pubkey       string `json:"pubkey"`
+		} `json:"message"`
+		Signature string `json:"signature"`
+	} `json:"entry"`
+}
+
+func (r *RemoteRelay) updateValidatorsMap(currentSlot uint64, retries int) error {
+	log.Info("requesting ", "currentSlot", currentSlot)
+	r.validatorsLock.Lock()
+	if r.validatorSyncOngoing {
+		r.validatorsLock.Unlock()
+		return errors.New("sync is ongoing")
+	}
+	r.validatorSyncOngoing = true
+	r.validatorsLock.Unlock()
+
+	newMap, err := r.getSlotValidatorMapFromRelay()
+	for err != nil && retries > 0 {
+		time.Sleep(time.Second)
```

Comment:
maybe log the error here?

Reply:
the flow here is a bit weird, that even if it's retried a bunch of times then gets out of the loop and proceeds through the function. error logging would be good, because this would be a critical problem.

Reply:
love that it retries, tho!